pkg/trie/triedb/codec/decode.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package codec
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+var (
+	ErrDecodeHashedStorageValue  = errors.New("cannot decode hashed storage value")
+	ErrDecodeHashedValueTooShort = errors.New("hashed storage value too short")
+	ErrReadChildrenBitmap        = errors.New("cannot read children bitmap")
+	// ErrDecodeChildHash is defined since no sentinel error is defined
+	// in the scale package.
+	ErrDecodeChildHash = errors.New("cannot decode child hash")
+	// ErrDecodeStorageValue is defined since no sentinel error is defined
+	// in the scale package.
+	ErrDecodeStorageValue = errors.New("cannot decode storage value")
+)
+
+const hashLength = common.HashLength
+
+// Decode decodes a node from a reader.
+// The encoding format is documented in the README.md
+// of this package, and specified in the Polkadot spec at
+// https://spec.polkadot.network/chap-state#defn-node-header
+func Decode(reader io.Reader) (n Node, err error) {
+	variant, partialKeyLength, err := decodeHeader(reader)
+	if err != nil {
+		return nil, fmt.Errorf("decoding header: %w", err)
+	}
+
+	switch variant {
+	case emptyVariant:
+		return Empty{}, nil
+	case leafVariant, leafWithHashedValueVariant:
+		n, err = decodeLeaf(reader, variant, partialKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode leaf: %w", err)
+		}
+		return n, nil
+	case branchVariant, branchWithValueVariant, branchWithHashedValueVariant:
+		n, err = decodeBranch(reader, variant, partialKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode branch: %w", err)
+		}
+		return n, nil
+	default:
+		// this is a programming error, an unknown node variant should be caught by decodeHeader.
+		panic(fmt.Sprintf("not implemented for node variant %08b", variant))
+	}
+}
+
+// decodeBranch reads from a reader and decodes to a node branch.
+// Note that we are not decoding the children nodes.
+func decodeBranch(reader io.Reader, variant variant, partialKeyLength uint16) (
+	node Branch, err error) {
+	node = Branch{}
+
+	node.PartialKey, err = decodeKey(reader, partialKeyLength)
+	if err != nil {
+		return Branch{}, fmt.Errorf("cannot decode key: %w", err)
+	}
```

Comment:
the `decodeKey` is common for `Branch` and `Leaf`, can we move this step to the `Decode` function and just pass the actual partial key to `decodeBranch/decodeKey`?

Reply:
Good catch! this will improve the nested error messages too!